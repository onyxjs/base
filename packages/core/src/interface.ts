import Suite, { isSuite, rootSymbol } from './suite';
import Test from './test';

const root = new Suite('root', { skip: false, todo: false }, null);
root[rootSymbol] = true;
let currentRoot: Suite = root;

export const getCurrentRoot: () => Suite = () => currentRoot;
export const setCurrentRoot: (v: Suite) => void = (v: Suite) => {
  if (!isSuite(v)) { return; }
  currentRoot = v;
};

export { root, currentRoot };

// istanbul ignore next internal
function _it(description: string, fn: () => void, skip = false, todo = false): Test {
  if (currentRoot.isRoot()) {
    throw new Error(`"${description}" "it" should not be called outside of "describe" block`);
  }

  const test = new Test(
    description,
    fn,
    {
      skip,
      todo,
    },
    currentRoot,
  );

  currentRoot.addChildren(test);
  return test;
}

export type ItFn = (description: string, fn: () => void) => Test;
// tslint:disable-next-line:interface-name
export interface It extends ItFn {
  skip: ItFn;
  todo: ItFn;
}

const it: It = ((description: string, fn: () => void) => _it(description, fn, false, false)) as It;
it.skip = (description: string, fn: () => void) => _it(description, fn, true, false);
it.todo = (description: string, fn: () => void) => _it(description, fn, false, true);

// istanbul ignore next internal
function _describe(description: string, fn: () => void, skip = false, todo = false): Suite {
  const suite = new Suite(
    description,
    {
      skip,
      todo,
    },
    currentRoot,
  );

  currentRoot = suite;
  fn();
  if (!suite.options.skip && !suite.options.todo && suite.children.length <= 0) {
    console.warn(`suite "${suite.getFullDescription()}" doesn't have any child tests or suites.`);
  }
  currentRoot = currentRoot.parent || root;

  currentRoot.addChildren(suite);
  return suite;
}

export type DescribeFn = (description: string, fn: () => void) => Suite;
// tslint:disable-next-line:interface-name
export interface Describe extends DescribeFn {
  skip: DescribeFn;
  todo: DescribeFn;
}

const describe: Describe = ((description: string, fn: () => void) =>
  _describe(description, fn, false, false)) as Describe;
describe.skip = (description: string, fn: () => void) => _describe(description, fn, true, false);
describe.todo = (description: string, fn: () => void) => _describe(description, fn, false, true);

export { it, describe };
export { beforeAll, afterAll, beforeEach, afterEach } from './hooks';
