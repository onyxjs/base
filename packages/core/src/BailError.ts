export class BailError extends Error {
  /* istanbul ignore next */
  constructor(message: string) {
    super(message)
    this.name = 'BailError'
  }
}
