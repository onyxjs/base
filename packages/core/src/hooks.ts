import { currentRoot } from './interface'

export type HookFn = (() => void) | (() => Promise<void>)
export type Hook = HookFn[]
export interface Hooks {
  beforeAll: Hook
  afterAll: Hook
  beforeEach: Hook
  afterEach: Hook
}
export type HookName = keyof Hooks

export const beforeAll: (...hooks: HookFn[]) => number = (...hooks: HookFn[]) => currentRoot.hooks.beforeAll.push(...hooks)
export const afterAll: (...hooks: HookFn[]) => number = (...hooks: HookFn[]) => currentRoot.hooks.afterAll.push(...hooks)
export const beforeEach: (...hooks: HookFn[]) => number = (...hooks: HookFn[]) => currentRoot.hooks.beforeEach.push(...hooks)
export const afterEach: (...hooks: HookFn[]) => number = (...hooks: HookFn[]) => currentRoot.hooks.afterEach.push(...hooks)
