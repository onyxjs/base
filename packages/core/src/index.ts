export {
  root,
  currentRoot,
  ItFn,
  It,
  it,
  DescribeFn,
  Describe,
  describe,
  getCurrentRoot,
  setCurrentRoot,
  beforeAll,
  afterAll,
  beforeEach,
  afterEach,
} from './interface'
export { Status, default as Result } from './result'
export { default as Runnable, isRunnable, RunnableTypes } from './runnable'
export { default as Test, isTest } from './test'
export { default as Suite, isSuite, SuiteStats, rootSymbol, BailError } from './suite'
export { default as Runner, RunOptions, runnerDefaults } from './runner'
export { Hook, HookName, HookFn, Hooks } from './hooks'
