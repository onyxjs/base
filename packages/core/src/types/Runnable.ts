export enum RunnableTypes {
  Runnable = 'runnable',
  Suite = 'suite',
  Test = 'test',
}

export interface RunnableOptions {
  skip: boolean
  todo: boolean
}
