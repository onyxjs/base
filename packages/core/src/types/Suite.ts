export interface SuiteStats {
  total: number
  pending: number
  running: number
  done: number
  passed: number
  failed: number
  skipped: number
  todo: number
  time: number
}
