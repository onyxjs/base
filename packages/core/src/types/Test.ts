export type TestFn = () => (void | Promise<any>)
