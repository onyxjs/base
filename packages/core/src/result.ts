// import { RunnableTypes } from "."\


// export enum Status {
//   Pending = 'pending',
//   Running = 'running',
//   Passed = 'passed',
//   Failed = 'failed',
//   Skipped = 'skipped',
//   Todo = 'todo',
// }

// Types
import { BailError } from './BailError'
import { TimeoutError } from './TimeoutError'
import { RunnableTypes, Status } from './types'

/**
 * @todo Delete messages.
 */
export default class Result {
  private _internalErrors: Error[]
  private _internalStatus: Status
  private _internalMessages: string[]
  
  public time: number
  public title: string
  public description: string
  public type: RunnableTypes

  constructor(messages: string | string[] = [], options: Result, errors: Error[] | Error = [], status?: Status) {
    this._internalErrors = !Array.isArray(errors) ? [ errors ] : errors
    this._internalStatus = status || Status.Pending
    this._internalMessages = !Array.isArray(messages) ? [ messages ] : messages
    
    this.description = options.description
    this.time = options.time
    this.title = options.title
    this.type = options.type
  }

  /**
   * @description Checks if the internal status is 'Pending' or 'Running'.
   */
  public isDone(): boolean {
    return this._internalStatus !== Status.Pending && this._internalStatus !== Status.Running
  }

  public get errors(): Array<Error> {
    return this._internalErrors
  }

  /**
   * @description Gets the internal status on the current `Result` instance.
   */
  public get status(): Status {
    return this._internalStatus
  }

  /**
   * @description Sets the internal status on the current `Result` instance.
   */
  public set status(v: Status) {
    if (this.isDone()) { return }
    this._internalStatus = v
  }

  /**
   * @description Gets the internal messages on the current `Result` instance.
   */
  public get messages(): string[] {
    return this._internalMessages
  }

  /**
   * @description Adds messages to the internal messages if the `Runnable` has not completed.
   */
  public addMessages(...messages: string[]): void {
    if (this.isDone()) { return }
    this._internalMessages = [...this._internalMessages, ...messages]
  }

  public addErrors(...errors: Error[]): void {
    if (this.isDone()) { return }
    this.addMessages(...errors.map((e) => e.message))
    this._internalErrors = [...this._internalErrors, ...errors]
  }
}
