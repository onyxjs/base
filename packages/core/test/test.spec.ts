import { Status } from '../src/result'
import Runnable from '../src/runnable'
import Suite from '../src/suite'
import Test, { isTest } from '../src/test'

describe('Test', () => {
  const defaultOpts = { skip: false, todo: false }

  it('should return isDone', async () => {
    const fn = jest.fn();
    const test = new Test('test', fn, defaultOpts, null);

    expect(test.isDone()).toBeFalsy();
    await test.run();
    expect(test.isDone()).toBeTruthy();
  });

  it('should check if is test', () => {
    const fn = jest.fn()

    expect(isTest(null)).toBeFalsy()
    expect(isTest({})).toBeFalsy()
    expect(isTest(new Runnable('not a test', defaultOpts, null))).toBeFalsy()
    expect(isTest(new Suite('not a test', defaultOpts, null))).toBeFalsy()
    expect(isTest(new Test('a test', fn, defaultOpts, null))).toBeTruthy()
  })

  it('should run', async () => {
    const fn = jest.fn()
    const test = new Test('test', fn, defaultOpts, null)

    const start = jest.fn()
    test.on('start', start)
    const pass = jest.fn()
    test.on('pass', pass)
    const end = jest.fn()
    test.on('end', end)

    expect(fn).toHaveBeenCalledTimes(0)
    expect(await test.run()).toMatchSnapshot()
    expect(fn).toHaveBeenCalledTimes(1)

    expect(start).toHaveBeenCalledTimes(1)
    expect(pass).toHaveBeenCalledTimes(1)
    expect(end).toHaveBeenCalledTimes(1)
  })

  it('should fail', async () => {
    const err = new Error('Fatal error')
    const fn = () => {
      throw err
    }
    const test = new Test('test', fn, defaultOpts, null)

    const start = jest.fn()
    test.on('start', start)
    const fail = jest.fn()
    test.on('fail', fail)
    const end = jest.fn()
    test.on('end', end)

    expect((await test.run()).status).toBe(Status.Failed)

    expect(start).toHaveBeenCalledTimes(1)
    expect(fail).toHaveBeenCalledTimes(1)
    expect(end).toHaveBeenCalledTimes(1)
  })

  it('should skip', async () => {
    const test = new Test('test', jest.fn(), { skip: true, todo: false}, null)

    expect((await test.run()).status).toBe('skipped')
  })

  it('should timeout', async () => {
    jest.useRealTimers();
    // Test fn should not resolve before the timeout promise
    const fn = () => new Promise((resolve) => {
        setTimeout(() => {
          resolve('Shouldn\'t resolve first');
        }, 3000);
      });

    const test = new Test('test', fn, defaultOpts, null);

    const end = jest.fn();
    test.on('end', end);
    const fail = jest.fn();
    test.on('fail', fail);
    const pass = jest.fn();
    test.on('pass', pass);
    const start = jest.fn();
    test.on('start', start);

    const result = await test.run({ timeout: 1000 });

    expect(result.status).toBe('failed');
    expect(result.messages[0]).toBe(`Error: ${test.description} has timed out: 1000ms`);
    expect(start).toHaveBeenCalledTimes(1);
    expect(fail).toHaveBeenCalledTimes(1);
    expect(end).toHaveBeenCalledTimes(1);
    expect(pass).toHaveBeenCalledTimes(0);
  });
});
