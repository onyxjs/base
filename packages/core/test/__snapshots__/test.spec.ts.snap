// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test should run 1`] = `
Result {
  "internalMessages": Array [],
  "internalStatus": "passed",
}
`;
